libkbfs/dir_data.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"github.com/keybase/client/go/logger"
+	"golang.org/x/net/context"
+)
+
+// dirBlockGetter is a function that gets a block suitable for
+// reading or writing, and also returns whether the block was already
+// dirty.  It may be called from new goroutines, and must handle any
+// required locks accordingly.
+type dirBlockGetter func(context.Context, KeyMetadata, BlockPointer,
+	path, blockReqType) (dblock *DirBlock, wasDirty bool, err error)
+
+// dirData is a helper struct for accessing and manipulating data
+// within a directory.  It's meant for use within a single scope, not
+// for long-term storage.  The caller must ensure goroutine-safety.
+type dirData struct {
+	dir    path
+	kmd    KeyMetadata
+	getter dirBlockGetter
+	cacher dirtyBlockCacher
+	log    logger.Logger
+	tree   *blockTree
+}
+
+func newDirData(dir path, kmd KeyMetadata, getter dirBlockGetter,
+	cacher dirtyBlockCacher, log logger.Logger) *dirData {
+	dd := &dirData{
+		dir:    dir,
+		kmd:    kmd,
+		getter: getter,
+		cacher: cacher,
+		log:    log,
+	}
+	dd.tree = &blockTree{
+		file:   dir,
+		kmd:    kmd,
+		getter: dd.blockGetter,
+	}
+	return dd
+}
+
+func (dd *dirData) rootBlockPointer() BlockPointer {
+	return dd.dir.tailPointer()
+}
+
+func (dd *dirData) blockGetter(
+	ctx context.Context, kmd KeyMetadata, ptr BlockPointer,
+	dir path, rtype blockReqType) (block Block, wasDirty bool, err error) {
+	return dd.getter(ctx, kmd, ptr, dir, rtype)
+}
+
+func (dd *dirData) getChildren(ctx context.Context) (
+	children map[string]EntryInfo, err error) {
+	topBlock, _, err := dd.getter(
+		ctx, dd.kmd, dd.rootBlockPointer(), dd.dir, blockRead)
+	if err != nil {
+		return nil, err
+	}
+
+	_, blocks, _, err := dd.tree.getBlocksForOffsetRange(
+		ctx, dd.rootBlockPointer(), topBlock, topBlock.FirstOffset(), nil,
+		false, true)
+	if err != nil {
+		return nil, err
+	}
+
+	numEntries := 0
+	for _, b := range blocks {
+		numEntries += len(b.(*DirBlock).Children)
```

Comment:
Is it worth two iterations to get the benefit of a single allocation of the slice?

Reply:
Actually yes!  I wrote a small program: https://play.golang.org/p/pZZG5V372W1

Executed on my machine, two loops is always faster than one loop:

```sh
$ ./x 10
One loop: 1.059µs
Two loops: 300ns
$ ./x 100
One loop: 13.799µs
Two loops: 426ns
$ ./x 1000
One loop: 20.654µs
Two loops: 7.353µs
$ ./x 10000
One loop: 384.259µs
Two loops: 71.895µs
$ ./x 100000
One loop: 2.366071ms
Two loops: 235.291µs
$ ./x 1000000
One loop: 14.520121ms
Two loops: 3.193464ms
$ ./x 10000000
One loop: 139.688947ms
Two loops: 27.569866ms
```

Reply:
THere's a third version :) https://play.golang.org/p/5l_xphR2NqZ

```
%  go run main.go 10000              ✘ 130
One loop: 346.775µs
One loop2: 86.786µs
Two loops: 115.841µs
```

Reply:
@songgao yeah, but that one doesn't apply to this code since we need to compute the new size from the data in the slice, we can't simply use the slice length.

Reply:
Ah that's right! I mis-read. Thanks.

Reply:
Great test, thanks!